common/src/gadgets/sw_cond_add.rs

```diff
@@ -9,42 +9,35 @@ use crate::domain::Domain;
 use crate::gadgets::booleanity::BitColumn;
 use crate::gadgets::cond_add::{CondAdd, CondAddValues};
 use crate::gadgets::{ProverGadget, VerifierGadget};
-use crate::{const_evals, AffineColumn, Column, FieldColumn};
-
-// Conditional affine addition:
-// if the bit is set for a point, add the point to the acc and store,
-// otherwise copy the acc value
-pub struct SwCondAdd<F: FftField, P: AffineRepr<BaseField = F>> {
-    pub(super) bitmask: BitColumn<F>,
-    pub(super) points: AffineColumn<F, P>,
-    // The polynomial `X - w^{n-1}` in the Lagrange basis
-    pub(super) not_last: FieldColumn<F>,
-    // Accumulates the (conditional) rolling sum of the points
-    pub acc: AffineColumn<F, P>,
-    pub result: P,
-}
+use crate::{const_evals, AffineColumn, Column};
+
+use super::cond_add::{AffineCondAdd, CondAddGen, CondAddValuesGen};
+
+pub type SWCondAdd<C> = CondAddGen<Affine<C>>;
+pub type SWCondAddValues<C> = CondAddValuesGen<Affine<C>>;
 
-pub struct SwCondAddValues<F: Field> {
-    pub bitmask: F,
-    pub points: (F, F),
-    pub not_last: F,
-    pub acc: (F, F),
+impl<C: SWCurveConfig> AffineCondAdd for Affine<C>
+where
+    <Self as AffineRepr>::BaseField: FftField,
+{
+    type CondAddT = SWCondAdd<C>;
 }
 
-impl<F, Curve> CondAdd<F, Affine<Curve>> for SwCondAdd<F, Affine<Curve>>
+impl<F, C> CondAdd<F, Affine<C>> for SWCondAdd<C>
 where
     F: FftField,
-    Curve: SWCurveConfig<BaseField = F>,
+    C: SWCurveConfig<BaseField = F>,
 {
-    type CondAddValT = SwCondAddValues<F>;
+    type Values = SWCondAddValues<C>;
+
     // Populates the acc column starting from the supplied seed (as 0 doesn't have an affine SW representation).
     // As the SW addition formula used is not complete, the seed must be selected in a way that would prevent
     // exceptional cases (doublings or adding the opposite point).
     // The last point of the input column is ignored, as adding it would made the acc column overflow due the initial point.
     fn init(
```

Comment:
Here you could check if the seed is outside of the group.

Reply:
This true. Sure. You could add  a comment saying you need to check PoP unless if S is outside of the group.